Websoket frame failes if the frame content is deflated partial EOM (0,0,-1,-1) from Tomcat.

Scenario is when the tomcat communicates to netty with websocket connection. The frmae size is 8192 bytes.
Tomcat sends the compressed websocket data frames to netty. 
When tomcat sends the websocket frames to netty, every frame ends with the EOM byte sequence which is 0,0,-1,-1. 
There are four conditions now:
1. The compressed data is small enough that the complete content and the EOM is contained in one frame. 

2. The data is large that it takes multiple frames to complete and the last frame contains some part data and the complete EOM. 

3. The  data is such that the EOM is in the two frames. e.g. the data is 8189 bytes, so the EOM starts in one and spreads over the next one. 

4.  The data and the EOM totals the 8192bytes. even in this case a next frame is sent with 0 as the first byte.

The conditions 1and 2 passes but 3 and 4 fails. Becuase the frame that contains only the partial EOM of previous frame is unable to be deflated in the DeflateDecoder
io.netty.handler.codec.compression.JdkZlibDecoder.decode(ChannelHandlerContext, ByteBuf, List<Object>)

### Expected behavior
The websocket message should be successfully sent. 

### Actual behavior
The message fils with the below exception 
Caused by: io.netty.handler.codec.CodecException: cannot read uncompressed buffer
at io.netty.handler.codec.http.websocketx.extensions.compression.DeflateDecoder.decode(DeflateDecoder.java:89)
at io.netty.handler.codec.http.websocketx.extensions.compression.PerMessageDeflateDecoder.decode(PerMessageDeflateDecoder.java:64)
at io.netty.handler.codec.http.websocketx.extensions.compression.PerMessageDeflateDecoder.decode(PerMessageDeflateDecoder.java:30)
at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:88)

### Steps to reproduce
Host a zip file in tomcat and transfer to netty application on a websocket connection.
But it is not reproducible with any zip file, only with the ones where content is not compressible by the java Deflater class. 

### Minimal yet complete reproducer code (or URL to code)
I have debugged in a complicated software, I am in process of creating a minimal code, but if the issue is crlear can we please get the fix. 

Though not in a perfect way, but I was able to fix the code by following changes in DeflateDecoder
[DeflateDecoder.txt](https://github.com/netty/netty/files/3828375/DeflateDecoder.txt)
:

boolean partialEOMFrame = true;
        for (int i = 0; i < msg.content().readableBytes(); i++) {
            byte contentAtI = msg.content().readByte();
            if (contentAtI != 0 && contentAtI != -1) {
                partialEOMFrame = false;
                break;
            }
        }
        msg.content().readerIndex(0);
;
;
;


if (!partialEOMFrame && readable && compositeUncompressedContent.numComponents() <= 0) {
            compositeUncompressedContent.release();
            throw new CodecException("cannot read uncompressed buffer");
        }


### Netty version
4.1.15
### JVM version (e.g. `java -version`)
1.8
### OS version (e.g. `uname -a`)
windows.

--- a/codec-http/src/main/java/io/netty/handler/codec/http/websocketx/extensions/compression/DeflateDecoder.java
+++ b/codec-http/src/main/java/io/netty/handler/codec/http/websocketx/extensions/compression/DeflateDecoder.java
@@ -136,8 +136,12 @@ abstract class DeflateDecoder extends WebSocketExtensionDecoder {
         // Correctly handle empty frames
         // See https://github.com/netty/netty/issues/4348
         if (!emptyDeflateBlock && readable && compositeDecompressedContent.numComponents() <= 0) {
-            compositeDecompressedContent.release();
-            throw new CodecException("cannot read uncompressed buffer");
+            // Sometimes after fragmentation the last frame
+            // May contain left-over data that doesn't affect decompression
+            if (!(msg instanceof ContinuationWebSocketFrame)) {
+                compositeDecompressedContent.release();
+                throw new CodecException("cannot read uncompressed buffer");
+            }
         }
 
         if (msg.isFinalFragment() && noContext) {

--- a/codec-http/src/test/java/io/netty/handler/codec/http/websocketx/extensions/compression/PerMessageDeflateDecoderTest.java
+++ b/codec-http/src/test/java/io/netty/handler/codec/http/websocketx/extensions/compression/PerMessageDeflateDecoderTest.java
@@ -328,4 +328,62 @@ public class PerMessageDeflateDecoderTest {
         assertFalse(decoderChannel.finish());
     }
 
+    @Test
+    public void testFragmentedFrameWithLeftOverInLastFragment() {
+        String hexDump = "677170647a777a737574656b707a787a6f6a7561756578756f6b7868616371716c657a6d64697479766d726f6" +
+                         "269746c6376777464776f6f72767a726f64667278676764687775786f6762766d776d706b76697773777a7072" +
+                         "6a6a737279707a7078697a6c69616d7461656d646278626d786f66666e686e776a7a7461746d7a776668776b6" +
+                         "f6f736e73746575637a6d727a7175707a6e74627578687871767771697a71766c64626d78726d6d7675756877" +
+                         "62667963626b687a726d676e646263776e67797264706d6c6863626577616967706a78636a72697464756e627" +
+                         "977616f79736475676f76736f7178746a7a7479626c64636b6b6778637768746c62";
+        EmbeddedChannel encoderChannel = new EmbeddedChannel(
+                ZlibCodecFactory.newZlibEncoder(ZlibWrapper.NONE, 9, 15, 8));
+        EmbeddedChannel decoderChannel = new EmbeddedChannel(new PerMessageDeflateDecoder(false));
+
+        ByteBuf originPayload = Unpooled.wrappedBuffer(ByteBufUtil.decodeHexDump(hexDump));
+        assertTrue(encoderChannel.writeOutbound(originPayload.duplicate().retain()));
+
+        ByteBuf compressedPayload = encoderChannel.readOutbound();
+        compressedPayload = compressedPayload.slice(0, compressedPayload.readableBytes() - 4);
+
+        int oneThird = compressedPayload.readableBytes() / 3;
+
+        TextWebSocketFrame compressedFrame1 = new TextWebSocketFrame(
+                false, WebSocketExtension.RSV1, compressedPayload.slice(0, oneThird));
+        ContinuationWebSocketFrame compressedFrame2 = new ContinuationWebSocketFrame(
+                false, WebSocketExtension.RSV3, compressedPayload.slice(oneThird, oneThird));
+        ContinuationWebSocketFrame compressedFrame3 = new ContinuationWebSocketFrame(
+                false, WebSocketExtension.RSV3, compressedPayload.slice(oneThird * 2, oneThird));
+        int offset = oneThird * 3;
+        ContinuationWebSocketFrame compressedFrameWithExtraData = new ContinuationWebSocketFrame(
+                true, WebSocketExtension.RSV3, compressedPayload.slice(offset,
+                     compressedPayload.readableBytes() - offset));
+
+        // check that last fragment contains only one extra byte
+        assertEquals(1, compressedFrameWithExtraData.content().readableBytes());
+        assertEquals(1, compressedFrameWithExtraData.content().getByte(0));
+
+        // write compressed frames
+        assertTrue(decoderChannel.writeInbound(compressedFrame1.retain()));
+        assertTrue(decoderChannel.writeInbound(compressedFrame2.retain()));
+        assertTrue(decoderChannel.writeInbound(compressedFrame3.retain()));
+        assertTrue(decoderChannel.writeInbound(compressedFrameWithExtraData));
+
+        // read uncompressed frames
+        TextWebSocketFrame uncompressedFrame1 = decoderChannel.readInbound();
+        ContinuationWebSocketFrame uncompressedFrame2 = decoderChannel.readInbound();
+        ContinuationWebSocketFrame uncompressedFrame3 = decoderChannel.readInbound();
+        ContinuationWebSocketFrame uncompressedExtraData = decoderChannel.readInbound();
+        assertFalse(uncompressedExtraData.content().isReadable());
+
+        ByteBuf uncompressedPayload = Unpooled.wrappedBuffer(uncompressedFrame1.content(), uncompressedFrame2.content(),
+                                      uncompressedFrame3.content(), uncompressedExtraData.content());
+        assertEquals(originPayload, uncompressedPayload);
+
+        assertTrue(originPayload.release());
+        assertTrue(uncompressedPayload.release());
+
+        assertTrue(encoderChannel.finishAndReleaseAll());
+        assertFalse(decoderChannel.finish());
+    }
 }

